Support Extended Modversions

maurer opened the pull request:
A companion to the [`EXTENDED_MODVERSIONS` kernel patch series](https://lore.kernel.org/lkml/20241123-extended-modversions-v10-0-0fa754ffdee3@google.com/), this PR adds support for `libkmod` to access the new modversions format when present, falling back to the original format otherwise.

Tested with `KDIR=` override selecting a kernel with `EXTENDED_MODVERSIONS=y` and `BASIC_MODVERSIONS=n` enabled with the series above in addition to a normal existing kernel.

maurer committed:
libkmod: Support EXTENDED_MODVERSIONS

In order to allow longer names, the kernel is taking a new MODVERSIONS
format option which stores the names in a strtab-like section and the
crcs in an array-like one. This patch makes `kmod_elf_get_modversions`
use either format, preferring the new one where present.

Signed-off-by: Matthew Maurer <mmaurer@google.com>
Link: https://github.com/kmod-project/kmod/pull/270
Signed-off-by: Lucas De Marchi <lucas.de.marchi@gmail.com>

maurer committed:
testsuite/test-modprobe: Test --show-modversions

We can't test actual CRC values, because they may vary depending on
kernel configuration. We can ensure the correct symbol names show up
though.

Signed-off-by: Matthew Maurer <mmaurer@google.com>
Link: https://github.com/kmod-project/kmod/pull/270
Signed-off-by: Lucas De Marchi <lucas.de.marchi@gmail.com>

stoeckmann commented:
Noticed these issues while reviewing changes: https://github.com/kmod-project/kmod/pull/272

Since they are already in master, I have created a new PR. No need to slow down functional review if there are bugs in the underlying base. ;)

These PRs should be kept in sync, depending which one is merged first.

lucasdemarchi reviewed (commented):
Overall looks good to me. I will try to play with them soon to make sure I didn't miss anything. I left some comments on each patch.

maurer committed:
libkmod: Centralize modversions parsing

Makes `kmod_elf_get_dependency_symbols` dispatch to
`kmod_elf_get_modversions` rather than open-coding the same version
extraction. This helps avoid duplicating logic, especially as we add
support for the extended format.

Link: https://github.com/kmod-project/kmod/pull/270
Signed-off-by: Lucas De Marchi <lucas.de.marchi@gmail.com>

maurer commented on libkmod/libkmod-elf.c:
```diff
@@ -42,6 +44,8 @@ static const char *const section_name_map[] = {
 	[KMOD_ELF_SECTION_SYMTAB] = ".symtab",
 	[KMOD_ELF_SECTION_VERSIONS] = "__versions",
+	[KMOD_ELF_SECTION_VERSION_NAMES] = "__version_ext_names",
+	[KMOD_ELF_SECTION_VERSION_CRCS] = "__version_ext_crcs",
```
Done

maurer commented on libkmod/libkmod-elf.c:
```diff
@@ -1111,37 +1181,26 @@ int kmod_elf_get_dependency_symbols(const struct kmod_elf *elf,
 		count++;
 
-		idx = kmod_elf_crc_find(elf, ver_off, versionslen, name, &crc);
+		idx = kmod_modversion_crc_find(version_array, vercount, name, &crc);
```
Done.

lucasdemarchi commented:
@maurer  the last commit with the additional testsuite patch breaks it because we don't have a way to test with that. I tried to replicate your env, but couldn't:

```
$ b4 shazam   20241123-extended-modversions-v10-0-0fa754ffdee3@google.com
Grabbing search results from lore.kernel.org
  Added from v11: 6 patches
  Added from v12: 6 patches
  Added from v13: 6 patches
...
Applying: module: get symbol crc back to unsigned
Patch failed at 0001 module: get symbol crc back to unsigned
```

maurer commented:
I'm not entirely sure why some of the targets aren't producing a module to run against. It looks like debian:unstable worked, but nothing else did. It's re-using an existing test module and just checking other properties on it, so this most likely means that the test is failing because kernel doesn't even have `CONFIG_MODVERSIONS` enabled, nothing to do with this feature. Do you have a preferred way to expect a test to go differently based on kernel configurations? The two ways I could see doing this are to either skip this test if `CONFIG_MODVERSIONS` is not enabled, or to run the test anyways, but if `CONFIG_MODVERSIONS` is not enabled, *expect* to fail to find version information.

As far as replicating my setup, the patches have landed upstream, so just grab a recent release and you'll see it, no need to `b4 shazam` any longer.

lucasdemarchi commented:
> CONFIG_MODVERSIONS

how would you check for it?  Running kernel may not be the tested kernel so it would require to check $KDIR/.config.

I think what we can do is to skip if the error is due to ENODATA instead of checking the config. We can do it easily with libkmod, but not easily with modprobe --dump-modversions.

> As far as replicating my setup, the patches have landed upstream, so just grab a recent release and you'll see it, no need to `b4 shazam` any longer.

ok, thanks.
